### PR TITLE
React CFButton component

### PR DESCRIPTION
**Pull request by dgreisen-cfpb**
fully tested

### CODE REVIEW

**Comment by Scotchester** [comment]
Gonna need some more info on this one :)

**Comment by Scotchester** [comment]
@dgreisen-cfpb  and I spoke by phone and agreed that we should try to get the coverage functionality working in Grunt or via @contolini's tool-less method before accepting this PR.

**Comment by imuchnik** [comment]
:+1:  I like it.

**Comment by ascott1** [comment]
This looks great @dgreisen-cfpb. :+1: from me.

**Comment by Scotchester** [comment]
:+1: The JS needs some linting, but I don't blame you for that, since the linter isn't in the code yet.

**Comment by dgreisen-cfpb** [comment]
@Scotchester clarified offline that linter is not blocker. So since it has unanimous approval, @Scotchester: want to do the honors and merge?

**Comment by Scotchester** [comment]
:shipit:

**Comment by Scotchester** [comment]
WAIT! Sorry. One more thing. We need to update the version number in `bower.json` and update the `CHANGELOG`.

**Comment by dgreisen-cfpb** [comment]
@Scotchester: done

**Comment by Scotchester** [comment]
:+1: Thanks!

**Comment by Scotchester** [comment]
@dgreisen-cfpb Can you please tag and release? Let me know if you have questions about that.

**Comment by cfarm** [comment]
@dgreisen-cfpb ditto @Scotchester! we're using this for OAH :)

**Comment by ascott1** [comment]
Nice @cfarm! Also check out https://github.com/ascott1/cf-icons/blob/react/src/CFIcon.jsx, which @dgreisen-cfpb and I were working on last week.

File: src/CFButton.jsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    rightIcon: React.PropTypes.element,
+    leftIcon: React.PropTypes.element,
+  },
+  getDefaultProps: function() {
+    return {
+      config: []
+    };
+  },
+  render: function() {
+    var rightIcon, leftIcon
```
Comment (anselmbradford) [inline src/CFButton.jsx:66]: Missing ending semicolon;

File: src/CFButton.jsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    var btnClassName = genClassName('btn', 'btn', this.props)
```
Comment (anselmbradford) [inline src/CFButton.jsx:68]: Missing ending semicolon;

File: src/CFButton.jsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  require('react/addons');
+}
+
+var allowedTypes = [
+  'super',
+  'disabled',
+  'warning',
+];
+
+function genClassName(requiredClasses, configPrefix, props) {
```
Comment (anselmbradford) [inline src/CFButton.jsx:12]: Consider underscore prefixing "private" methods localized to only this module: `_genClassName`.

File: src/CFButton.jsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }).join('')
+  var className = requiredClasses + configClasses
+  if (props.className) {
+    className += ' ' + props.className;
+  }
+  return className
+}
+
+var CFButton = React.createClass({
+  /*
```
Comment (anselmbradford) [inline src/CFButton.jsx:25]: Use jsdoc comments http://usejsdoc.org/

File: src/CFButton.jsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+    if (this.props.leftIcon) {
+      leftIcon = React.addons.cloneWithProps(this.props.leftIcon, {className: 'btn_icon__left'});
+    }
+
+    var buttonProps = omit(this.props, 'config', 'rightIcon', 'leftIcon', 'className', 'children');
+
+    if (this.props.href) {
+      return <a className={btnClassName} {...buttonProps}>{leftIcon}{this.props.children}{rightIcon}</a>
+    } else {
```
Comment (anselmbradford) [inline src/CFButton.jsx:81]: else statement is unneeded.

File: src/CFButton.jsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      return <button className={btnClassName} {...buttonProps}>{leftIcon}{this.props.children}{rightIcon}</button>
+    }
+  }
+});
+
+function omit(obj) {
+  var omitted = Array.prototype.slice.call(arguments, 1);
```
Comment (anselmbradford) [inline src/CFButton.jsx:88]: See warning on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments

Comment (anselmbradford) [inline src/CFButton.jsx:88]: May not matter due to the length being sliced.

File: src/CFButton.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  var out = {};
+  var props = Object.getOwnPropertyNames(obj);
+  for (var i=0;i<props.length;i++) {
+    var prop = props[i];
+    if (omitted.indexOf(prop) < 0) {
+      out[prop] = obj[prop];
+    }
+  }
+  return out
```
Comment (anselmbradford) [inline src/CFButton.jsx:97]: Missing semicolon;

File: src/CFButton.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      return <button className={btnClassName} {...buttonProps}>{leftIcon}{this.props.children}{rightIcon}</button>
+    }
+  }
+});
+
+function omit(obj) {
+  var omitted = Array.prototype.slice.call(arguments, 1);
+  var out = {};
+  var props = Object.getOwnPropertyNames(obj);
+  for (var i=0;i<props.length;i++) {
```
Comment (anselmbradford) [inline src/CFButton.jsx:91]: Length lookup each loop can be prevented with:

``` js
for ( var i = 0, len = props.length; i < len; i++ ) {
```

File: src/CFButton.jsx (hunk 3 above)
Comment (dgreisen-cfpb) [inline src/CFButton.jsx:12]: this will (hopefully) soon be pulled out into a helper repo.

**Comment by dgreisen-cfpb** [comment]
@Scotchester: I do not know how to tag and release. Probably not the best person to do that, either.

**Pull request merged**